enhancements/storage/vsphere-configurable-maximum-allowed-number-of-block-volumes-per-node.md

```diff
@@ -0,0 +1,237 @@
+
+5. Documentation:
+
+   - Update the vSphere CSI driver documentation to include information about the new feature and how to configure it.
+   - Include a statement informing users of the current requirement of having a homogeneous cluster with all nodes
+     running ESXi 8 or higher. Until this requirement is met, the limit set in maxAllowedBlockVolumesPerNode must not be
+     increased to a higher value than 59.
+   - Point out that setting the value to `0` will disable the limit.
+
+### Workflow Description
+
+1. Administrator configures the limit:
+   - The administrator creates or updates a ClusterCSIDriver object to specify the desired maximum number of volumes per
+     node using the new `maxAllowedBlockVolumesPerNode` API field.
+2. Operator reads configuration:
+   - The vSphere CSI Operator monitors the ClusterCSIDriver object for changes.
+   - Upon detecting a change, the operator reads the configured limit value.
+3. Operator updates the new limit for DaemonSet:
+   - The operator updates the DaemonSet for the vSphere CSI driver, injecting the `MAX_VOLUMES_PER_NODE` environment
+     variable with the configured limit value into the driver node pods on worker nodes.
+4. Driver reflects the limit in deployments:
+   - The vSphere CSI driver checks that `max-pvscsi-targets-per-vm` FSS is set to true and reflects the limit as
+     `MAX_VOLUMES_PER_NODE` environment variable and uses the configured limit during
+     volume provisioning requests.
+
+### API Extensions
+
+- New field in ClusterCSIDriver CRD:
+  - A new CRD field will be introduced to represent the maximum volume limit configuration.
+  - This CRD will contain a single new field (e.g., `spec.driverConfig.vSphere.maxAllowedBlockVolumesPerNode`) to define
+    the desired limit.
+  - The field will default to the current maximum limit of 59 volumes per node for vSphere 7.
+  - The CRD will validate the value fits within the defined range (0-255).
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+No unique considerations for Hypershift. The configuration and behavior of the vSphere CSI driver with respect to the
+maximum volume limit will remain consistent across standalone and managed clusters.
+
+#### Standalone Clusters
+
+This enhancement is fully applicable to standalone OpenShift clusters.
+
+#### Single-node Deployments or MicroShift
+
+No unique considerations for MicroShift. The configuration and behavior of the vSphere CSI driver with respect to the
+maximum volume limit will remain consistent across standalone and SNO/MicroShift clusters.
+
+### Implementation Details/Notes/Constraints
+
+One of the possible future constraints might be increasing the limit with newer vSphere versions. However, we expect the
+limit to be increasing rather than decreasing and making the API validation more relaxed is possible.
+
+### Risks and Mitigations
+
+1. Unlimited volume attachments per node:
```

Comment:
We could also block users from setting this value to `0` via some kind of validation in CRD.

Reply:
Absolutely, more on this in API PR: https://github.com/openshift/api/pull/2190#discussion_r1944522981